Activate Ruby environment using version managers

vinistock opened the pull request:
Closes #21

Add Ruby environment activation through version managers. This PR
- Exposes a new configuration to select a version manager
- Implements how to activate the Ruby environment based on each manager
- Starts activating the environment before trying to run the `rdbg` executable

ko1 commented:
Sorry I forget to response on it.

(1) do you plan to make a npm package for it?
(2) For "none" I think it is enough to use `injectRubyEnvironment` without `rbenv`, ... is enough (and remove all other shell dependent code), at least on rbenv. What do you think about it? In other words, I don't want to ask setting `rdbg.rubyVersionManager` if not needed.

ko1 commented:
Another question.

On rbenv, `RBENV_VERSION` is specified at first setting. if `.ruby-version` is rewritten, it seems it doesn't applied for VSCode without restarting the VSCode process. Is it intentional? I'm not sure how often such a case happened.

vinistock commented:
I've rebased the PR and modified the changes in #124, since we want all shells to use interactive mode to load `rbenv`, `chruby` or any other manager.

Concerning your questions
1. Yes, we hope to turn the activation into a package so that we don't have to keep duplicating code between the Ruby LSP, RDBG and any other extension that wants to launch Ruby
2. I'm not sure I understand this one. When the version manager is `none` (which is the default), we don't invoke `injectRubyEnvironment`
3. If you want to re-activate the Ruby environment and inject the variables into the node process when `.ruby-version` changes, then we have to use a VS Code file watcher to track changes to the file and then trigger re-activation if any modifications happen

ko1 commented:
Thank you for your quick reply.

> Yes, we hope to turn the activation into a package so that we don't have to keep duplicating code between the Ruby LSP, RDBG and any other extension that wants to launch Ruby

If it will be released soon, I want to wait for it.

> I'm not sure I understand this one. When the version manager is none (which is the default), we don't invoke injectRubyEnvironment

Yes. My idea is "assuming users use `rbenv` and `injectRubyEnvironment` for `none` will setup `rbenv` environment variables and no need to call shell for the following commands. But it has an issue for (3).

> If you want to re-activate the Ruby environment and inject the variables into the node process when .ruby-version changes, then we have to use a VS Code file watcher to track changes to the file and then trigger re-activation if any modifications happen

I understand the technique. My question was it is intentional or not for `.ruby-version` changes. As I wrote " I'm not sure how often such a case happened." so if the real case is not many, I agree we can ignore it.

ono-max commented on src/extension.ts:
```diff
@@ -272,20 +294,68 @@ class RdbgAdapterDescriptorFactory implements DebugAdapterDescriptorFactory {
 
 	make_shell_command(cmd: string) {
 		const shell = process.env.SHELL;
-		switch (true) {
-			case shell && (shell.endsWith("bash") || shell.endsWith("fish")):
-				return shell + " -l -c '" + cmd + "'";
-			case shell && shell.endsWith("zsh"):
-				// As the recommended way, initialization commands for rbenv are written in ".zshrc".
-				// However, it's not loaded on the non-interactive shell.
-				// Thus, we need to run this command as the interactive shell.
-				// FYI: https://zsh.sourceforge.io/Guide/zshguide02.html
-				return shell + " -l -c -i '" + cmd + "'";
-			default:
-				return cmd;
+
+		if (this.support_login(shell)) {
+			return shell + " -lic '" + cmd + "'";
+		} else {
+			return cmd;
```
Can you cut out this part to another PR? I can merge this part soon.

vinistock replied:
Why not merge this PR instead?

ono-max replied:
Sorry, I can't merge this PR since @ko1-san is reviewing this PR, and he decides it. And the reason why I ask you to cut this part to another PR is that it is significant to merge this part regardless of this feature.

ono-max replied:
But, it's just a suggestion, I'll leave it up to you.

vinistock replied:
I understand your point of view, but my sincere opinion is that we should focus on getting this PR merged.

Using Ruby version managers is common practice on the community and making sure the environment is activated properly is essential to get all of the benefits of RDBG in the editor.

ko1 commented:
I think no configuration is better but I feel current proposed documentation asks user to write it if they uses a version manager. So I want to rewrite it after merge.
@vinistock could you allow me to modify then after merging?

vinistock commented:
Can you clarify what you mean by `no configuration`? Do you mean trying to guess which version manager the developer uses?

Totally okay with you refactoring as needed. Just trying to understand your perspective since the long term vision is for the Ruby LSP and RDBG to both use an NPM package for Ruby environment activation.

vinistock commented:
We're adding the `auto` option as the [default for the Ruby LSP](https://github.com/Shopify/vscode-ruby-lsp/pull/437), which automatically detects the user's version manager. Would that what you're thinking about?

ko1 commented:
@vinistock 
Sorry for the late response but it seems conflicting.
Could you rebase it or can I make a patch?

firien commented:
`promisify` was removed with #41. I have nothing against `promisify` - it was really the embedded `require` that caught my eye and provided impetus for e1040ba.
For consistency, either this should adopt the Promise wrapped callback style, or #41 should revert to `promisify` version. As noted in #41 - there are also event based versions of `child_process.exec` in here - so this would introduce yet another flavor of `child_process.exec`

ono-max commented:
Thanks @firien san.

I'm ok with using `promisify`, so let's use `promisify` in https://github.com/ruby/vscode-rdbg/pull/41, in another PR.
